src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -112,141 +175,261 @@ abstract class LazyModule()(implicit val p: Parameters)
+    *
+    * @param bc [[LazyModule]] instance that needs to be wrapped.
+    * @param valName [[ValName]] used to name this instance,
+    *                it can be automatically generated by [[ValName]] macro, or specify manually.
+    * @param sourceInfo [[SourceInfo]] information about where this [[LazyModule]] is being generated
+    */
   def apply[T <: LazyModule](bc: T)(implicit valName: ValName, sourceInfo: SourceInfo): T = {
-    // Make sure the user put LazyModule around modules in the correct order
-    // If this require fails, probably some grandchild was missing a LazyModule
-    // ... or you applied LazyModule twice
-    require (scope.isDefined, s"LazyModule() applied to ${bc.name} twice ${sourceLine(sourceInfo)}")
+    // Make sure the user puts [[LazyModule]] around modules in the correct order.
+    require (scope.isDefined,
+             s"LazyModule() applied to ${bc.name} twice ${sourceLine(sourceInfo)}. Ensure that descendant LazyModules are instantiated with the LazyModule() wrapper and that you did not call LazyModule() twice.")
     require (scope.get eq bc, s"LazyModule() applied to ${bc.name} before ${scope.get.name} ${sourceLine(sourceInfo)}")
+    // [[LazyModule.scope]] stack pop.
     scope = bc.parent
     bc.info = sourceInfo
     if (!bc.suggestedNameVar.isDefined) bc.suggestName(valName.name)
     bc
   }
 }
 
+/** trait which must be returned in [[LazyModule.module]] for generating actual hardware. */
 sealed trait LazyModuleImpLike extends RawModule
 {
+  /** The [[LazyModule]] which will generate this [[LazyModuleImpLike]]. */
   val wrapper: LazyModule
+  /** The IOs that will be automatically "punched" for this [[LazyModuleImpLike]]  */
   val auto: AutoBundle
+  /** The metatdata that describes the [[HalfEdge]]s which generated [[auto]]. */
   protected[diplomacy] val dangles: Seq[Dangle]
 
-  // .module had better not be accessed while LazyModules are still being built!
-  require (!LazyModule.scope.isDefined, s"${wrapper.name}.module was constructed before LazyModule() was run on ${LazyModule.scope.get.name}")
+  // [[wrapper.module]] had better not be accessed while LazyModules are still being built!
+  require (LazyModule.scope.isEmpty, s"${wrapper.name}.module was constructed before LazyModule() was run on ${LazyModule.scope.get.name}")
 
+  /** Set module name. */
   override def desiredName = wrapper.desiredName
   suggestName(wrapper.suggestedName)
 
+  /** [[Parameters]] for chisel [[Module]]s. */
   implicit val p = wrapper.p
 
+  /** [[instantiate]] will be called when a instance of [[LazyModuleImp]] is called. */
   protected[diplomacy] def instantiate() = {
     val childDangles = wrapper.children.reverse.flatMap { c =>
       implicit val sourceInfo = c.info
+      // Calling [[c.module]] will push the real [[Module]] into [[chisel3.internal.Builder]].
+      //  Note: this place is not calling [[instantiate]], it is just pushing [[c.module]] to Builder.
```

Comment:
I think we might have an excessive amount of comments in this function. It seems that every statement has a fairly long comment, which starts to impair the readability of the actual code itself. I'd suggest removing most of them, and writing a slightly more detailed description for this method to describe the high level points of the algorithm.

Reply:
I agree. I am not sure HOW this algorithm is doing what I think it is doing though... It's basically looking for any edges that you haven't manually connected(bound) already  within the module instantiation, and then either connecting them for you (internally) or "auto-punching" ports onto the Module if the edge is actually leaving this module.

Reply:
Sure I will get it as simple as possible. I will write a summary comments for `instantiate`, rather than inline comments.

Reply:
i realized that i was wrong, there is no binding to I/Os anymore. I am going to mark this conversation as resolved.